### PR TITLE
Revert "kserve: add grpconly envoy filter (#888)"

### PR DESCRIPTION
**Pull request by dtrifiro**
This reverts commit 48e0b8bad9bbae951731a36c6e17bf26981fed7b (#888).

This was a temporary fix for https://issues.redhat.com/browse/RHOAIENG-165 which does not affect caikit+tgis, tgis or ovms anymore.

testing instructions:
1. check out https://github.com/opendatahub-io/opendatahub-operator/pull/888 and spin it up
2. check that the kserve-temporary-fixes feature has been added and that grpc-only EnvoyFilter exists.
3. Spin this PR up
4. check that the kserve-temporary-fixes feature has been deleted and that the grpc-only EnvoyFilter has been deleted

### CODE REVIEW

**Comment by israel-hdez** [comment]
How / where is the formal fix done?

**Comment by bartoszmajsak** [comment]
I would also like to address another thing as part of this "revert".

How do we make sure this filter is removed from the already running clusters? 

We should have a logic to handle such cases.  Before we come up with something more advanced I think we can simply add some logic to the bootstrap process (`main.go`) and remove `FeatureTracker` by name  `appNamespace + "-" + name` ("opendatahub-kserve-temporary-fixes"). Though for that we have to expose FT `func`s to be usable from `main.go`. 

WDYT @aslakknutsen @zdtsw ?

**Comment by zdtsw** [comment]
> How / where is the formal fix done?

same question:  is a fix already merged somewhere in upstream? or how/why need a revert here

**Comment by zdtsw** [comment]
> I would also like to address another thing as part of this "revert".
> 
> How do we make sure this filter is removed from the already running clusters?
> 
> We should have a logic to handle such cases. Before we come up with something more advanced I think we can simply add some logic to the bootstrap process (`main.go`) and remove `FeatureTracker` by name `appNamespace + "-" + name` ("opendatahub-kserve-temporary-fixes"). Though for that we have to expose FT `func`s to be usable from `main.go`.
> 
> WDYT @aslakknutsen @zdtsw ?

will that be suitable to catch all the cases (e.g remove unuse resource, old Featuretracker, temp. workaround) into `upgrade.go` than in `main.go` (or have a new file only handle these "cleanup" only)

**Comment by dtrifiro** [comment]
@israel-hdez it was an upstream fix in the python `grpcio` module. Updating this dependency in `tgis` and `caikit`, fixed the issue,

**Comment by bartoszmajsak** [comment]
> > I would also like to address another thing as part of this "revert".
> > How do we make sure this filter is removed from the already running clusters?
> > We should have a logic to handle such cases. Before we come up with something more advanced I think we can simply add some logic to the bootstrap process (`main.go`) and remove `FeatureTracker` by name `appNamespace + "-" + name` ("opendatahub-kserve-temporary-fixes"). Though for that we have to expose FT `func`s to be usable from `main.go`.
> > WDYT @aslakknutsen @zdtsw ?
> 
> will that be suitable to catch all the cases (e.g remove unuse resource, old Featuretracker, temp. workaround) into `upgrade.go` than in `main.go` (or have a new file only handle these "cleanup" only)

I wouldn't say it captures all cases, as some `Feature`s can have custom cleanup logic, but for this given one it would suffice.  I need to think it through.

**Comment by israel-hdez** [comment]
> @israel-hdez it was an upstream fix in the python `grpcio` module. Updating this dependency in `text-generation-inference` and `caikit` fixed the issue.

@dtrifiro You should, then, get in touch with dashboard team and coordinate this PR to go together with an update to the images. I still see older images here: https://github.com/opendatahub-io/odh-dashboard/blob/4b8b9b80b725914635e1d593ccf3f22969a799d3/manifests/modelserving/kustomization.yaml.

Also, please follow-up Bartosz and Wen feedback:

> How do we make sure this filter is removed from the already running clusters? We should have a logic to handle such cases. Before we come up with something more advanced I think we can simply [...] remove FeatureTracker.

**Comment by dtrifiro** [comment]
Regarding the manifests: the manifests you linked include grpcio 1.62, which is not affected by RHOAIENG-165, meaning this is out already.

Regarding the resources cleanup: working on a solution, starting from `upgrade.go`

**Comment by israel-hdez** [comment]
> Regarding the manifests: the manifests you linked include grpcio 1.62, which is not affected by RHOAIENG-165, meaning this is out already.
> 

It calls my attention that the manifests are from March 1st, while the workaround PR #888 was opened March 4th.

> Regarding the resources cleanup: working on a solution, starting from `upgrade.go`


Thanks for following-up.

**Comment by dtrifiro** [comment]
One corner I have: odh 2.9 has been released, but RHOAI 2.9 has been dropped. Do we really need to do a proper cleanup of the envoyfilter? The temporary fix has never been deployed on RHOAI for sure.

Regarding the cleanup: I have a naive implementation in `upgrade.go` that deletes the enovyfilter by selecting  he kind, instance name and namespace, but I haven't been so far able to do something similar using the `FeatureTracker` API as discussed.

@zdtsw @bartoszmajsak

**Comment by zdtsw** [comment]
> One corner I have: odh 2.9 has been released, but RHOAI 2.9 has been dropped. Do we really need to do a proper cleanup of the envoyfilter? The temporary fix has never been deployed on RHOAI for sure.
> 
> Regarding the cleanup: I have a naive implementation in `upgrade.go` that deletes the enovyfilter by selecting he kind, instance name and namespace, but I haven't been so far able to do something similar using the `FeatureTracker` API as discussed.
> 
> @zdtsw @bartoszmajsak

we no need to get the first "workaround" out for downstream, aka i think RHOAIENG-165  should be closed instead.

as for the "cleanup" we will still need it for ODH.

File: pkg/upgrade/upgrade.go
Comment (israel-hdez) [inline pkg/upgrade/upgrade.go:377]: It would be nicer if you use our own types: https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/apis/features/v1/groupversion_info.go#L30

Something like `v1.GroupVersion.WithKind("FeatureTracker")` (assuming the right `v1` is imported).

File: pkg/upgrade/upgrade.go
Comment (israel-hdez) [inline pkg/upgrade/upgrade.go:361]: Just for sanity check? I see no references...

File: pkg/upgrade/upgrade.go
Comment (israel-hdez) [inline pkg/upgrade/upgrade.go:371]: @bartoszmajsak would the `opendatahub-` prefix change? is there a const for it?

**Review by israel-hdez — COMMENTED** [review]
The outcome seems to be the needed one (notice that I didn't try the code).

I'll let the operator owners to decide if this is good, since I see there is also the `CleanupExistingResource` which seems to have a similar goal.

File: pkg/upgrade/upgrade.go
Comment (dtrifiro) [reply pkg/upgrade/upgrade.go]: Fixed so that it uses `dscApplicationsNamespace`

File: pkg/upgrade/upgrade.go
Comment (dtrifiro) [reply pkg/upgrade/upgrade.go:239]: Yes, it's a sanity check, following the scheme of other similar functions. I'm assuming it could be removed, since we try to get the `FeatureTracker` by name and ignore NotFound errors (assuming trying to fetch a non-existing crd returns a NotFound too)

File: pkg/upgrade/upgrade.go
Comment (cam-garrison) [inline pkg/upgrade/upgrade.go:258]: Don't believe it's necessary to print the error here since it looks like it gets logged in the call above in the chain if not nil.

File: pkg/upgrade/upgrade.go
Comment (zdtsw) [inline pkg/upgrade/upgrade.go:248]: maybe i did not get the whole picture here, but why not just call  removeFeatureTracker() or even Cleanup()?

File: pkg/upgrade/upgrade.go
Comment (dtrifiro) [reply pkg/upgrade/upgrade.go]: Agreed, deleted in the last push

File: pkg/upgrade/upgrade.go
Comment (dtrifiro) [reply pkg/upgrade/upgrade.go]: Cleanup sounds like a good way of doing it. Thanks for the feedback, I was having trouble trying to figure out the `Feature`/`FeatureTracker` APIs

**Comment by bartoszmajsak** [comment]
@dtrifiro I failed to review this today, but I put it as the first thing for my morning queue.

**Comment by bartoszmajsak** [comment]
@dtrifiro I think this PR is a great candidate to be part of #957 work which is ongoing. We need a simplified approach for removing resources we don't need anymore. As this filter is not part of the RHOAI release, but only upstream release so far, if we do not include #888 in the coming downstream release, we can refine your solution and leverage #957. 

@zdtsw @VaishnaviHire would this work? (not including #888 in the downstream release)

**Comment by zdtsw** [comment]
> @dtrifiro I think this PR is a great candidate to be part of #957 work which is ongoing. We need a simplified approach for removing resources we don't need anymore. As this filter is not part of the RHOAI release, but only upstream release so far, if we do not include #888 in the coming downstream release, we can refine your solution and leverage #957.
> 
> @zdtsw @VaishnaviHire would this work? (not including #888 in the downstream release)

correct, this PR only need to be in ODH not need downstream.
we can use the same function from #957 once it is merged and get this done in ODH 2.11.0 release which is planned in another 2 weeks for codefreeze

**Comment by zdtsw** [comment]
should we finalize this PR before ODH 2.12.0 code freeze? (10th May)

**Comment by dtrifiro** [comment]
@zdtsw #957  hasn't been merged yet and I haven't had the bandwidth to work on this, I'm afraid it will have to wait some more time

**Comment by zdtsw** [comment]
> @zdtsw #957 hasn't been merged yet and I haven't had the bandwidth to work on this, I'm afraid it will have to wait some more time

no worries, lets leave it for odh 2.13.0

**Comment by dtrifiro** [comment]
Is there anything holding this back?

**Comment by zdtsw** [comment]
@israel-hdez can you do a final review and we can have it finalized.

**Comment by bartoszmajsak** [comment]
@dtrifiro could you add steps to test it in the PR description? I would like to give it a spin.

**Comment by dtrifiro** [comment]
@bartoszmajsak 

1. check out https://github.com/opendatahub-io/opendatahub-operator/pull/888 and spin it up
2. check that the `kserve-temporary-fixes` feature has been added and that `grpc-only` `EnvoyFilter` exists.
3. Spin this PR up
4. check that the `kserve-temporary-fixes` feature has been deleted and that the `grpc-only` `EnvoyFilter` has been deleted

File: pkg/upgrade/upgrade.go
Comment (bartoszmajsak) [inline pkg/upgrade/upgrade.go:261]: As I mentioned a [while back](https://github.com/opendatahub-io/opendatahub-operator/pull/931#issuecomment-2014464759) the name of the `FeatureTracker`, as it's a cluster-wide resource, it's composed from appNs and feature name, so without this change the cleanup actually does not work.

```suggestion
	deprecatedFeatureTrackers := []string{dscApplicationsNamespace + "-kserve-temporary-fixes"}
```

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
Without this change the cleanup actually does not work

File: pkg/upgrade/upgrade.go
Comment (zdtsw) [reply pkg/upgrade/upgrade.go:261]: @dtrifiro if we can get this piece fixed? it need application namespace as prefix.
we can try have this done by tomorrow in to odh 2.15 release

Comment (dtrifiro) [reply pkg/upgrade/upgrade.go]: @bartoszmajsak @zdtsw done

**Comment by zdtsw** [comment]
LGTM

**Review by cam-garrison — APPROVED** [review]
lgtm with app namespace change now added to FT name

**Pull request merged**
